### ISSUE TITLE
Please read this if you want a working version on NC 27

### ISSUE THREAD

**Issue by Xefir** [e-alfred/epubreader#49]
Hello.

I rewrite a good part of the app on my own repository here : https://git.crystalyx.net/Xefir/epubreader
You can download the latest release [here](https://git.crystalyx.net/Xefir/epubreader/releases) and unzip/untar the archive or clone the repository in your apps directory to get the new working version.

On top of making the app works with the latest version of Nextcloud, I've done some other things like : 
- Typing everything to have a robust and easy development for the future
- Follow the [Nextcloud Coding Standard](https://github.com/nextcloud/coding-standard)
- Add [PSALM](https://psalm.dev/) to have a strict static analysis of the code and detect code mistakes
- Move the `vendor` folder to `public` because `vendor` is usually used by composer and not for storing publicly JS assets
- Add a development environment via composer

I won't provide a PR for this or publish my work on the Nextcloud AppStore, for a bunch of reasons : 
- There are way too much modifications to review it properly
- There is already 3 different apps for epubreader on the store, and I don't want to become this [xkcd](https://xkcd.com/927/)
- Moving the `vendor` folder seems to have broken the [nextcloud transifex bot](https://github.com/e-alfred/epubreader/commits?author=nextcloud-bot) and I don't known how I can fix it (and it won't work without Github anyway)
- I don't like hosting sources on Github when I can have it at home, that's why we use Nextcloud isn't it ? :D

But I won't make this app mine and I don't want to.
Do not hesitate to grab my work, fork it and put it back on Github if you want !
I will be glad if anyone (and maybe @e-alfred too !) want to continue my work !

Thank you for reading this and have a good time reading your precious books :)

**Comment by benjaminfrombe** [comment]
thanks for your work ! unfortunately I can't get it to work. After enabling I also am unable to go to my admin page

`[index] Error: Exception: Class "OCA\Epubreader\AppInfo\Application" not found in file '/var/www/html/custom_apps/epubreader/lib/Settings/Personal.php' line 57 at <<closure>>

0. /var/www/html/lib/private/AppFramework/App.php line 183
   OC\AppFramework\Http\Dispatcher->dispatch(["OCA\\Settings\ ... "], "index")
1. /var/www/html/lib/private/Route/Router.php line 315
   OC\AppFramework\App::main("OCA\\Settings\\ ... r", "index", ["OC\\AppFramewo ... "], ["overview","set ... "])
2. /var/www/html/lib/base.php line 1064
   OC\Route\Router->match("/settings/admin/overview")
3. /var/www/html/index.php line 36
   OC::handleRequest()

Caused by:

Error: Class "OCA\Epubreader\AppInfo\Application" not found at <<closure>>

 0. /var/www/html/lib/private/Settings/Manager.php line 220
    OCA\Epubreader\Settings\Personal->getSection()
 1. /var/www/html/lib/private/Settings/Manager.php line 334
    OC\Settings\Manager->getSettings("personal", "additional")
 2. /var/www/html/lib/private/Settings/Manager.php line 295
    OC\Settings\Manager->getPersonalSettings("additional")
 3. /var/www/html/apps/settings/lib/Controller/CommonSettingsTrait.php line 111
    OC\Settings\Manager->getPersonalSections()
 4. /var/www/html/apps/settings/lib/Controller/CommonSettingsTrait.php line 62
    OCA\Settings\Controller\AdminSettingsController->formatPersonalSections("admin", "overview")
 5. /var/www/html/apps/settings/lib/Controller/CommonSettingsTrait.php line 148
    OCA\Settings\Controller\AdminSettingsController->getNavigationParameters("admin", "overview")
 6. /var/www/html/apps/settings/lib/Controller/AdminSettingsController.php line 68
    OCA\Settings\Controller\AdminSettingsController->getIndexResponse("admin", "overview")
 7. /var/www/html/lib/private/AppFramework/Http/Dispatcher.php line 230
    OCA\Settings\Controller\AdminSettingsController->index("overview")
 8. /var/www/html/lib/private/AppFramework/Http/Dispatcher.php line 137
    OC\AppFramework\Http\Dispatcher->executeController(["OCA\\Settings\ ... "], "index")
 9. /var/www/html/lib/private/AppFramework/App.php line 183
    OC\AppFramework\Http\Dispatcher->dispatch(["OCA\\Settings\ ... "], "index")
10. /var/www/html/lib/private/Route/Router.php line 315
    OC\AppFramework\App::main("OCA\\Settings\\ ... r", "index", ["OC\\AppFramewo ... "], ["overview","set ... "])
11. /var/www/html/lib/base.php line 1064
    OC\Route\Router->match("/settings/admin/overview")
12. /var/www/html/index.php line 36
    OC::handleRequest()

GET /settings/admin/overview
from 172.18.0.1 by admin at 2023-06-21T09:06:34+00:00`

**Comment by Xefir** [comment]
Hello @benjaminfrombe 

Can you provide me your Nextcloud and PHP versions please ?

**Comment by ummon29** [comment]
Hello,

after a quick test this  (1.4.9) runs fine for me with Nextcloud 26.0.3 and PHP 8.1.
Thanks for the great work. I'm happy to have a usable version of this App again :-)

**Comment by benjaminfrombe** [comment]
Hi !

I'm using the docker latest version. => Nextcloud 27.0.0

docker exec -it -u www-data nextcloud php -v

PHP 8.2.7 (cli) (built: Jun 14 2023 05:26:32) (NTS)

Copyright (c) The PHP Group

Zend Engine v4.2.7, Copyright (c) Zend Technologies

    with Zend OPcache v8.2.7, Copyright (c), by Zend Technologies


thanks,


Benjamin

**Comment by devnoname120** [comment]
@Xefir Ah I just heard about your fork. Too bad because I already forked the project: https://github.com/devnoname120/epubviewer

I'm not sure yet how actively I will maintain `epubviewer` but I may take a look at your repo. Publishing the app in the NextCloud App Store is important because not all users have access to the filesystem/configuration of the server/VPS on which their instance runs.

For example [Hetzner Storage Share](https://www.hetzner.com/storage/storage-share) (managed NextCloud) only allows you to install apps from the App Store, not import your own.

**Comment by Xefir** [comment]
I push an update that fix a bug I introduce on folder sharing (and refacto again to have a even nicer stack).

@devnoname120 I see what you did and I understand that not all users will use my fork, but that's not my goal either.
In fact, if we can take over the apps that is already on the store (or remove them), I'll be happy to push my fork on it.
But renaming **again** like you did will ultimately make another app appear in it and the end users won't install any of them, it's just frustrating.

I'm working on another Nextcloud project that take all my time and I work on the epubreader fork when I have time.
I have finished to clean the PHP part, this part is done for me.
But the JS part is ... Erk ... Ugly as fuck and need to be redo from scratch IMHO.
Maybe one day ...

**Comment by devnoname120** [comment]
@Xefir I agree with your sentiment that it's annoying to have so many outdated forks on the store. However I would rather have n+1 forks with 1 that works, rather than n forks with 0 that works. Having it on the store is important for me because my NextCloud instance is managed by Hetzner and I can't sideload any apps, they must be installed from the store.

Nonetheless it should be possible to remove (or redirect) the older entries from the store by contacting NextCloud people. Ultimately I would like to [move my repository to the NextCloud organization](https://docs.nextcloud.com/server/latest/developer_manual/app_publishing_maintenance/publishing.html#moving-your-repo-to-the-nextcloud-organization) so that ownership can be more easily transferred in the future without creating yet another fork.

**Comment by Xefir** [comment]
@devnoname120 If you app is moved to a official Nextcloud repository, I'll be glad to make you a Pull Request with ALL my changes.
It'll be very annoying for you to review, but I prefer to have an official fork settled once and for all like you say.

Take me in touch when the repository is ready :)

**Comment by devnoname120** [comment]
@Xefir Soumds good to me. Note that it may be easier to just grab your changes, push them to my repo, publish a new release, and only *then* apply for moving the repo to the NextCloud organization. It should be easier to have them accept a code base that is closer to their standards.

**Comment by devnoname120** [comment]
@Xefir git.crystalyx.net is down right now. Do you have a mirror?

**Comment by devnoname120** [comment]
@Xefir Don't hesitate to contribute to https://github.com/devnoname120/epubviewer ! Definitely need some help here :)

**Comment by Xefir** [comment]
@devnoname120 Will do. I work on other apps for now but I would like to give back some time when I can ^^"
